lib/msal-core/src/AuthError.ts

```diff
@@ -0,0 +1,240 @@
+        Object.setPrototypeOf(this, AuthError.prototype);
+    }
+    static createUnexpectedError(errDesc: string) {
+        return new AuthError(ErrorMessage.unexpectedError.code, errDesc);
+    }
+}
+
+/**
+ * @hidden
+ *
+ * Error thrown when there is an error in the client code running on the browser.
+ */
+export class ClientAuthError extends AuthError {
+    constructor(error: string, errorDesc: string) {
+        super(error, errorDesc);
+        this.name = "ClientAuthError";
+
+        Object.setPrototypeOf(this, ClientAuthError.prototype);
+    }
+
+    static createEndpointResolutionError(errDesc: string): ClientAuthError {
+        var errorMessage = "Error: could not resolve endpoints. Please check network and try again.";
+        if (!Utils.isEmpty(errDesc)) {
+            errorMessage += " Details: " + errDesc;
+        }
+        return new ClientAuthError(ErrorMessage.endpointResolutionError.code, errorMessage);
+    }
+
+    static createMultipleMatchingTokensInCacheError(scope: string): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.multipleMatchingTokens.code,
+            "Cache error for scope " + scope + ": " + ErrorMessage.multipleMatchingTokens.desc);
+    }
+
+    static createMultipleAuthoritiesInCacheError(scope: string): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.multipleMatchingAuthorities.code,
+            "Cache error for scope " + scope + ": " + ErrorMessage.multipleMatchingAuthorities.desc);
+    }
+
+    static createPopupWindowError(): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.popUpWindowError.code,
+            ErrorMessage.popUpWindowError.desc);
+    }
+
+    static createTokenRenewalTimeoutError(): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.tokenRenewalError.code,
+            ErrorMessage.tokenRenewalError.desc);
+    }
+
+    static createInvalidStateError(invalidState: string, actualState: string): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.invalidStateError.code,
+            ErrorMessage.invalidStateError.desc + invalidState + ", state expected : " + actualState);
+    }
+
+    static createNonceMismatchError(invalidNonce: string, actualNonce: string): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.nonceMismatchError.code,
+            ErrorMessage.nonceMismatchError + invalidNonce + ", nonce expected : " + actualNonce);
+    }
+
+    static createLoginInProgressError(): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.loginProgressError.code,
+            ErrorMessage.loginProgressError.desc);
+    }
+
+    static createAcquireTokenInProgressError(): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.acquireTokenProgressError.code,
+            ErrorMessage.acquireTokenProgressError.desc);
+    }
+
+    static createUserCancelledError(): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.userCancelledError.code,
```

Comment:
I do appreciate the config of the errors quite a bit, makes it easy to change and reference, but I wonder if there is a way to get that but also get readability of errors in code.

When I am getting a SOMETHING_ERROR it is nice to be able to easily trace that and see oh here is where that error is being created, instead of ending up in a config file and looking around.

its more of a dev nicety, and I am not sure I have a great answer for an alternative off the top of my head, but food for thought

Reply:
The error.stack should have the trace where it occurred in the code

Reply:
Is there something additional beyond a verbal description of the error and the stack trace that we should be giving back?